File: iosApp/Shared/Screens/Calendar/CalendarScreen.swift
//
//  CalendarScreen.swift
//  Well
//
//  Created by Phil on 01.12.2021.
//  Copyright © 2021 Well. All rights reserved.
//

import SwiftUI
import SharedMobile

private typealias Feature = CalendarFeature

struct CalendarScreen: View {
    let state: CalendarFeature.State
    let listener: (CalendarFeature.Msg) -> Void

    @State
    private var dialogMeeting: MeetingViewModel?

    var body: some View {
        VStack(spacing: 0) {
            CalendarMonthView(
                state: state.infoState,
                title: {
                    CalendarTitleView(state: state, showNextMonth: showNextMonth, showPrevMonth: showPrevMonth)
                },
                colors: CalendarMonthViewColors.companion.mainCalendar,
                onSelect: {
                    listener(Feature.MsgCompanion.shared.SelectDate(selectedDate: $0.date))
                },
                showNextMonth: showNextMonth,
                showPrevMonth: showPrevMonth
            )
            MeetingsList(
                meetings: state.selectedItemMeetings.map { [ $0] } ?? state.upcomingMeetings,
                onSelectMeeting: {
                    dialogMeeting = $0
                },
                onDeleteMeeting: {
                    listener(Feature.MsgUpdateMeetingState(meeting: $0, state: $1))
                },
                onStartCall: {
                    listener(Feature.MsgStartCall(meeting: $0))
                },
                onSelectUser: {
                    listener(Feature.MsgOpenUserProfile(meeting: $0))
                },
				onDeleteRequest: {
					listener(Feature.MsgDeleteRequest(meeting: $0))
				},
				onUpdateState: {
					listener(Feature.MsgUpdateMeetingState(meeting: $0, state: $1))
				}
			)
        }.padding(.horizontal)
            .fillMaxSize()
            .background(GradientView(gradient: .main).ignoresSafeArea(edges: .top))
    }

    func showNextMonth() {
        listener(Feature.MsgCompanion.shared.NextMonth)
    }

    func showPrevMonth() {
        listener(Feature.MsgCompanion.shared.PrevMonth)
    }
}

private struct CalendarTitleView: View {
    let state: Feature.State

    let showNextMonth: () -> Void
    let showPrevMonth: () -> Void

    @State
    private var currentMonthNameSize: CGSize?
    private let titleId = "titleId"

    var body: some View {
        VStack {
            ZStack {
                GeometryReader { geometry in
                    ScrollViewReader { scrollView in
                        ScrollView(.horizontal) {
                            HStack(spacing: 25) {
                                HStack {
                                    Spacer()
                                    otherMonthTitle(state.prevMonthName, singleLetter: false)
                                }.fillMaxWidth().frame(minWidth: geometry.size.width / 2)
                                currentMonthText
                                    .id(titleId)
                                    .sizeReader {
                                        currentMonthNameSize = $0
                                    }
                                HStack {
                                    otherMonthTitle(state.nextMonthName, singleLetter: false)
                                    Spacer()
                                }.fillMaxWidth().frame(minWidth: geometry.size.width / 2)
                            }
                            .onAppear {
                                scrollView.scrollTo(titleId, anchor: .center)
                            }
                            .onChange(of: state.currentMonthName) { _ in
                                scrollView.scrollTo(titleId, anchor: .center)
                            }
                        }
                    }
                }.frame(height: currentMonthNameSize?.height)
                HStack(spacing: 0) {
                    Button(action: showPrevMonth) {
                        otherMonthTitle(state.prevMonthName, singleLetter: true)
                            .fillMaxWidth()
                    }
                    currentMonthText
                        .frame(size: currentMonthNameSize)
                    Button(action: showNextMonth) {
                        otherMonthTitle(state.nextMonthName, singleLetter: true)
                            .fillMaxWidth()
                    }
                }.background(Color.black.opacity(0.01)).opacity(0.1)
            }
            Spacer().frame(height: 18)
            Text(state.year.toString())
                .textStyle(.body2)
        }.foregroundColor(.white).padding(.vertical, 30)
    }

    var currentMonthText: some View {
        Text(state.currentMonthName)
            .textStyle(.h3)
    }

    func otherMonthTitle(_ title: String, singleLetter: Bool) -> some View {
        Text(
            singleLetter ?
            title.first?.toString() ?? ""
            : title
        )
        .opacity(0.2)
        .textStyle(.h5)
    }
}

private struct MeetingsList: View {
    let meetings: [Feature.StateDayMeetings]
    let onSelectMeeting: (MeetingViewModel) -> Void
    let onDeleteMeeting: (MeetingViewModel, Meeting.StateCanceled) -> Void
    let onStartCall: (MeetingViewModel) -> Void
    let onSelectUser: (MeetingViewModel) -> Void
	let onDeleteRequest: (MeetingViewModel) -> Void
    let onUpdateState: (MeetingViewModel, Meeting.State) -> Void

    var body: some View {
        ScrollView(showsIndicators: false) {
            LazyVStack(spacing: 15) {
                ForEach(meetings, id: \.day) { dayMeetings in
                    Text(dayMeetings.day.localizedRelatedDescription())
                        .textStyle(.body1)
                        .foregroundColor(.white)
                    ForEach(dayMeetings.meetings) { meeting in
                        if meeting.waitingExpertResolution {
                            RequestedMeetingCard(
                                meeting: meeting,
                                onSelectUser: {
                                    onSelectUser(meeting)
                                },
                                onUpdateState: {
                                    onUpdateState(meeting, $0)
                                }
                            )
                        } else {
                            ConfirmedMeetingCard(
                                meeting: meeting,
                                onSelect: {
                                    onSelectMeeting(meeting)
                                },
                                onStartCall: {
                                    onStartCall(meeting)
                                },
                                onSelectUser: {
                                    onSelectUser(meeting)
                                },
								onDeleteRequest: {
									onDeleteRequest(meeting)
								},
                                onDelete: { state in
                                    onDeleteMeeting(meeting, state)
								}
                            )
                        }
                    }
                }
            }.padding(.vertical, 15)
        }
    }
}

private struct ConfirmedMeetingCard: View {
    let meeting: MeetingViewModel

    let onSelect: () -> Void
    let onStartCall: () -> Void
    let onSelectUser: () -> Void
	let onDeleteRequest: () -> Void
    let onDelete: (Meeting.StateCanceled) -> Void

    @State
    private var deletionReasonDialogShowing = false

    var body: some View {
        let isOngoing = meeting.status == .ongoing
        ZStack(alignment: isOngoing ? .trailing : .topTrailing) {
            HStack {
                VStack(alignment: .leading, spacing: 0) {
                    let ongoing = meeting.status == .ongoing
                    Text(ongoing ? Feature.Strings.shared.now : meeting.startTime.toString())
                        .strikethrough(meeting.status == .past)
                        .foregroundColorKMM(ongoing ? .companion.Green : .companion.Black)
                        .textStyle(.caption)

                    Text(meeting.title)
                        .textStyle(.subtitle2)
                        .foregroundColor(.black)
                        .padding(.top, 3)
                        .padding(.bottom, meeting.rejectionReason == nil ? 9 : 0)

                    meeting.rejectionReason.map {
                        Text($0)
                            .textStyle(.subtitle2)
                            .foregroundColor(.black)
                            .padding(.top, 3)
                            .padding(.bottom, 9)
                    }

                    Button(action: onSelectUser) {
                        HStack {
                            ProfileImage(meeting.otherUser)
                                .frame(size: 20)
                            Spacer().frame(width: 7)
                            Text("with \(meeting.otherUser.fullName)")
                                .textStyle(.caption)
                                .foregroundColorKMM(.companion.LightBlue)
                        }
                    }
                }
            }.fillMaxWidth(alignment: .leading).padding(15).padding(.leading, 12)
            if isOngoing {
                Button(action: onStartCall) {
                    Image(systemName: "phone.fill")
                        .font(.system(size: 20))
                        .foregroundColorKMM(.companion.Green)
                        .padding()
                }
            } else {
                Button {
                    if meeting.rejectionReason != nil {
                        onDelete(Meeting.StateCanceled(reason: ""))
                    } else {
                        deletionReasonDialogShowing = true
                    }
                } label: {
                    Image(systemName: "trash.fill")
                        .font(.system(size: 20))
                        .foregroundColorKMM(.companion.RadicalRed)
                        .padding()
                }
                .textFieldAlertController(
                    isPresented: $deletionReasonDialogShowing,
                    info: .init(
                        title: Feature.Strings.shared.deletionTitle,
                        placeholder: Feature.Strings.shared.deletionLabel
                    ),
                    alertConfirmed: {
                        onDelete(Meeting.StateCanceled(reason: $0))
                    }
                )
            }
        }
        .background(Shapes.medium.foregroundColor(.white))
        .onTapGesture(perform: onSelect)
    }
}

private struct RequestedMeetingCard: View {
    let meeting: MeetingViewModel

    let onSelectUser: () -> Void
    let onUpdateState: (Meeting.State) -> Void

    @State
    private var rejectionReasonDialogShowing = false

    var body: some View {
        HStack {
            VStack(alignment: .leading, spacing: 10) {
                if meeting.isExpert {
                    HStack {
                        Button(action: onSelectUser) {
                            Text(meeting.otherUser.fullName)
                                .textStyle(.subtitle2)
                                .foregroundColorKMM(.companion.Green)
                        }
                        Text(Feature.Strings.shared.needsYourHelp)
                            .textStyle(.subtitle2)
                    }
                } else {
                    Text(meeting.title)
                        .textStyle(.subtitle2)
                        .foregroundColor(.black)
                }
                Text("\(Feature.Strings.shared.bookingTime): \(meeting.startTime)")
                    .textStyle(.caption)
                HStack {
                    Button(action: {
                        onUpdateState(Meeting.StateConfirmed.shared)
                    }) {
                        Text(Feature.Strings.shared.confirm)
                            .textStyle(.subtitle2)
                            .foregroundColorKMM(.companion.MediumBlue)
                    }
                    Button(action: {
                        rejectionReasonDialogShowing = true
                    }) {
                        Text(Feature.Strings.shared.reject)
                            .textStyle(.subtitle2)
                            .foregroundColorKMM(.companion.Pink)
                    }
                    .textFieldAlertController(
                        isPresented: $rejectionReasonDialogShowing,
                        info: .init(
                            title: Feature.Strings.shared.rejectionTitle,
                            placeholder: Feature.Strings.shared.rejectionLabel
                        ),
                        alertConfirmed: {
                            onUpdateState(Meeting.StateRejected(reason: $0))
                        }
                    )
                }
            }.layoutPriority(1)
            Spacer().fillMaxWidth()
        }.padding(15).padding(.leading, 12)
            .background(Shapes.medium.foregroundColor(.white))
    }
}

File: modules/androidUi/src/main/kotlin/com/well/modules/androidUi/composableScreens/calendar/CalendarScreen.kt
package com.well.modules.androidUi.composableScreens.calendar

import com.well.modules.androidUi.customViews.CalendarMonthView
import com.well.modules.androidUi.customViews.CalendarTitleScope
import com.well.modules.androidUi.customViews.ProfileImage
import com.well.modules.androidUi.customViews.gradientBackground
import com.well.modules.androidUi.ext.minus
import com.well.modules.androidUi.ext.plus
import com.well.modules.androidUi.ext.start
import com.well.modules.androidUi.ext.toColor
import com.well.modules.features.calendar.calendarFeature.CalendarFeature.Msg
import com.well.modules.features.calendar.calendarFeature.CalendarFeature.State
import com.well.modules.models.Color
import com.well.modules.models.Meeting
import com.well.modules.models.MeetingViewModel
import com.well.modules.models.date.dateTime.localizedRelatedDescription
import com.well.modules.utils.kotlinUtils.ifTrueOrNull
import com.well.modules.utils.viewUtils.CalendarMonthViewColors
import com.well.modules.utils.viewUtils.Gradient
import com.well.modules.features.calendar.calendarFeature.CalendarFeature as Feature
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.ButtonDefaults
import androidx.compose.material.Card
import androidx.compose.material.Icon
import androidx.compose.material.IconButton
import androidx.compose.material.LocalContentColor
import androidx.compose.material.MaterialTheme
import androidx.compose.material.ProvideTextStyle
import androidx.compose.material.Text
import androidx.compose.material.TextButton
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Call
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.layout
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp

@Composable
fun CalendarScreen(
    state: State,
    listener: (Msg) -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .gradientBackground(Gradient.Main)
            .padding(horizontal = 15.dp)
            .statusBarsPadding()
    ) {
        CalendarMonthView(
            state = state.infoState,
            title = { TitleView(state) },
            colors = CalendarMonthViewColors.mainCalendar,
            onSelect = {
                listener(Msg.SelectDate(it))
            },
            showNextMonth = {
                listener(Msg.NextMonth)
            },
            showPrevMonth = {
                listener(Msg.PrevMonth)
            },
        )
        val (dialogMeeting, setDialogMeeting) = remember { mutableStateOf<MeetingViewModel?>(null) }
        MeetingsList(
            state.selectedItemMeetings?.let(::listOf)
                ?: state.upcomingMeetings,
            onSelectMeeting = setDialogMeeting,
            onStartCall = {
                listener(Msg.StartCall(it))
            },
            onSelectUser = {
                listener(Msg.OpenUserProfile(it))
            },
            onUpdateState = { meeting, state ->
                listener(Msg.UpdateMeetingState(meeting, state))
            },
        )
    }
}

@Composable
private fun CalendarTitleScope.TitleView(state: State) {
    CompositionLocalProvider(LocalContentColor provides Color.White.toColor()) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier
                .padding(vertical = 30.dp)
        ) {
            Row {
                TitleOtherMonth(
                    text = state.prevMonthName,
                    moveToLeft = true,
                    onClick = ::showPrevMonth,
                )
                Text(
                    state.currentMonthName,
                    style = MaterialTheme.typography.h3,
                    modifier = Modifier.padding(horizontal = 25.dp)
                )
                TitleOtherMonth(
                    text = state.nextMonthName,
                    moveToLeft = false,
                    onClick = ::showNextMonth,
                )
            }
            Spacer(Modifier.height(18.dp))
            Text(
                state.year.toString(),
                style = MaterialTheme.typography.body2,
            )
        }
    }
}

@Composable
private fun RowScope.TitleOtherMonth(
    text: String,
    moveToLeft: Boolean,
    onClick: () -> Unit,
) {
    TextButton(
        onClick = onClick,
        colors = ButtonDefaults.textButtonColors(
            contentColor = Color.White.toColor().copy(alpha = 0.2f)
        ),
        modifier = Modifier
            .weight(1f)
            .layout { measurable, constraints ->
                val placeable = measurable.measure(constraints.copy(maxWidth = Int.MAX_VALUE))
                val width = minOf(placeable.width, constraints.maxWidth)
                layout(width, placeable.height) {
                    placeable.place(
                        x = if (moveToLeft) width - placeable.width else 0,
                        y = 0
                    )
                }
            }
    ) {
        Text(
            text,
            style = MaterialTheme.typography.h5,
        )
    }
}

@Composable
private fun MeetingsList(
    meetings: List<State.DayMeetings>,
    onSelectMeeting: (MeetingViewModel) -> Unit,
    onStartCall: (MeetingViewModel) -> Unit,
    onSelectUser: (MeetingViewModel) -> Unit,
    onUpdateState: (MeetingViewModel, Meeting.State) -> Unit,
) {
    LazyColumn(
        verticalArrangement = Arrangement.spacedBy(15.dp),
        contentPadding = PaddingValues(vertical = 15.dp)
    ) {
        meetings.forEach { dayMeetings ->
            item(key = dayMeetings.day.toString()) {
                Box(
                    contentAlignment = Alignment.Center,
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Text(
                        dayMeetings.day.localizedRelatedDescription(),
                        style = MaterialTheme.typography.body1,
                        color = Color.White.toColor()
                    )
                }
            }
            items(dayMeetings.meetings, key = { it.id.toString() }) { meeting ->
                if (meeting.waitingExpertResolution) {
                    RequestedMeetingCard(
                        meeting = meeting,
                        onUpdateState = {
                            onUpdateState(meeting, it)
                        },
                        onSelectUser = {
                            onSelectUser(meeting)
                        },
                    )
                } else {
                    ConfirmedMeetingCard(
                        meeting = meeting,
                        onSelect = {
                            onSelectMeeting(meeting)
                        },
                        onStartCall = {
                            onStartCall(meeting)
                        },
                        onSelectUser = {
                            onSelectUser(meeting)
                        },
                    )
                }
            }
        }
    }
}

@Composable
private fun ConfirmedMeetingCard(
    meeting: MeetingViewModel,
    onSelect: () -> Unit,
    onSelectUser: () -> Unit,
    onStartCall: () -> Unit,
) {
    Card(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(
                enabled = meeting.status == MeetingViewModel.Status.Upcoming,
                onClick = onSelect,
            )
    ) {
        Box(
            Modifier
                .padding(15.dp)
                .padding(start = 12.dp)
        ) {
            Column {
                ProvideTextStyle(value = MaterialTheme.typography.caption) {
                    if (meeting.status == MeetingViewModel.Status.Ongoing) {
                        Text(
                            Feature.Strings.now,
                            color = Color.Green.toColor(),
                        )
                    } else {
                        Text(
                            meeting.startTime.toString(),
                            textDecoration = ifTrueOrNull(meeting.status == MeetingViewModel.Status.Past) {
                                TextDecoration.LineThrough
                            },
                            color = Color.Black.toColor(),
                        )
                    }
                }
                Text(
                    meeting.title,
                    style = MaterialTheme.typography.subtitle2,
                    color = Color.Black.toColor(),
                    modifier = Modifier.padding(top = 3.dp, bottom = 9.dp)
                )
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier
                        .clip(CircleShape)
                        .clickable {
                            onSelectUser()
                        }
                ) {
                    ProfileImage(
                        user = meeting.otherUser,
                        modifier = Modifier
                            .padding(end = 7.dp)
                            .size(20.dp)
                    )
                    Text(
                        "with ${meeting.otherUser.fullName}",
                        style = MaterialTheme.typography.caption,
                        color = Color.LightBlue.toColor()
                    )
                }
            }
            if (meeting.status == MeetingViewModel.Status.Ongoing) {
                IconButton(
                    onClick = onStartCall,
                    modifier = Modifier
                        .align(Alignment.CenterEnd)
                ) {
                    Icon(
                        Icons.Rounded.Call,
                        contentDescription = null,
                        tint = Color.Green.toColor(),
                    )
                }
            }
        }
    }
}

@Composable
private fun RequestedMeetingCard(
    meeting: MeetingViewModel,
    onSelectUser: () -> Unit,
    onUpdateState: (Meeting.State) -> Unit,
) {
    Card(
        modifier = Modifier
            .fillMaxWidth()
    ) {
        Column(
            modifier = Modifier
                .padding(
                    PaddingValues(15.dp) + PaddingValues(start = 12.dp) - ButtonDefaults.TextButtonContentPadding
                )
                .fillMaxWidth()
        ) {
            if (meeting.isExpert) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    TextButton(onClick = onSelectUser) {
                        Text(
                            meeting.otherUser.fullName,
                            style = MaterialTheme.typography.subtitle2,
                        )
                    }
                    Text(
                        Feature.Strings.needsYourHelp,
                        style = MaterialTheme.typography.subtitle2,
                    )
                }
            } else {
                Text(
                    meeting.title,
                    style = MaterialTheme.typography.subtitle2,
                )
            }
            Text(
                "${Feature.Strings.bookingTime}: ${meeting.startTime}",
                style = MaterialTheme.typography.caption,
                modifier = Modifier.padding(
                    start = ButtonDefaults.TextButtonContentPadding.start()
                )
            )

            Row {
                TextButton(onClick = {
                    onUpdateState(Meeting.State.Confirmed)
                }) {
                    Text(
                        Feature.Strings.confirm,
                        style = MaterialTheme.typography.subtitle2,
                        color = Color.MediumBlue.toColor(),
                    )
                }
                Spacer(Modifier.width(10.dp))
                TextButton(onClick = {
                    onUpdateState(Meeting.State.Rejected)
                }) {
                    Text(
                        Feature.Strings.reject,
                        style = MaterialTheme.typography.subtitle2,
                        color = Color.Pink.toColor(),
                    )
                }
            }
        }
    }
}